File: pc/protocol.go
package main

import (
	"encoding/binary"
	"fmt"
	"io"
)

// Message types
const (
	MsgVideoInfo   byte = 0x01
	MsgVideoFrame  byte = 0x02
	MsgAudioInfo   byte = 0x03
	MsgAudioPacket byte = 0x04
	MsgTouchEvent  byte = 0x05
	MsgHeartbeat   byte = 0x06
)

// Touch actions
const (
	TouchDown byte = 0x00
	TouchMove byte = 0x01
	TouchUp   byte = 0x02
)

// VideoInfo payload
type VideoInfo struct {
	Width  uint16
	Height uint16
	FPS    uint8
}

// AudioInfo payload
type AudioInfo struct {
	SampleRate uint32
	Channels   uint8
	Bits       uint8
}

// TouchEvent payload
type TouchEvent struct {
	X         uint16
	Y         uint16
	Action    uint8
	PointerID uint8
}

// TimestampedData holds a nanosecond timestamp + raw data
type TimestampedData struct {
	Timestamp int64  // nanoseconds since stream start
	Data      []byte
}

// EncodeVideoInfo serializes video info
func EncodeVideoInfo(v VideoInfo) []byte {
	buf := make([]byte, 5)
	binary.BigEndian.PutUint16(buf[0:], v.Width)
	binary.BigEndian.PutUint16(buf[2:], v.Height)
	buf[4] = v.FPS
	return buf
}

// DecodeVideoInfo deserializes video info
func DecodeVideoInfo(data []byte) (VideoInfo, error) {
	if len(data) < 5 {
		return VideoInfo{}, fmt.Errorf("video info too short")
	}
	return VideoInfo{
		Width:  binary.BigEndian.Uint16(data[0:]),
		Height: binary.BigEndian.Uint16(data[2:]),
		FPS:    data[4],
	}, nil
}

// EncodeAudioInfo serializes audio info
func EncodeAudioInfo(a AudioInfo) []byte {
	buf := make([]byte, 6)
	binary.BigEndian.PutUint32(buf[0:], a.SampleRate)
	buf[4] = a.Channels
	buf[5] = a.Bits
	return buf
}

// DecodeAudioInfo deserializes audio info
func DecodeAudioInfo(data []byte) (AudioInfo, error) {
	if len(data) < 6 {
		return AudioInfo{}, fmt.Errorf("audio info too short")
	}
	return AudioInfo{
		SampleRate: binary.BigEndian.Uint32(data[0:]),
		Channels:   data[4],
		Bits:       data[5],
	}, nil
}

// EncodeTouchEvent serializes touch event
func EncodeTouchEvent(t TouchEvent) []byte {
	buf := make([]byte, 6)
	binary.BigEndian.PutUint16(buf[0:], t.X)
	binary.BigEndian.PutUint16(buf[2:], t.Y)
	buf[4] = t.Action
	buf[5] = t.PointerID
	return buf
}

// DecodeTouchEvent deserializes touch event
func DecodeTouchEvent(data []byte) (TouchEvent, error) {
	if len(data) < 6 {
		return TouchEvent{}, fmt.Errorf("touch event too short")
	}
	return TouchEvent{
		X:         binary.BigEndian.Uint16(data[0:]),
		Y:         binary.BigEndian.Uint16(data[2:]),
		Action:    data[4],
		PointerID: data[5],
	}, nil
}

// PrependTimestamp prepends an 8-byte big-endian nanosecond timestamp to data
func PrependTimestamp(data []byte, ts int64) []byte {
	buf := make([]byte, 8+len(data))
	binary.BigEndian.PutUint64(buf[0:], uint64(ts))
	copy(buf[8:], data)
	return buf
}

// ExtractTimestamp extracts an 8-byte timestamp from the beginning of data
func ExtractTimestamp(data []byte) (int64, []byte, error) {
	if len(data) < 8 {
		return 0, nil, fmt.Errorf("data too short for timestamp")
	}
	ts := int64(binary.BigEndian.Uint64(data[0:]))
	return ts, data[8:], nil
}

// WriteMessage writes a framed message to the writer
func WriteMessage(w io.Writer, msgType byte, payload []byte) error {
	frame := make([]byte, 5+len(payload))
	frame[0] = msgType
	binary.BigEndian.PutUint32(frame[1:], uint32(len(payload)))
	copy(frame[5:], payload)
	_, err := w.Write(frame)
	return err
}

// ReadMessage reads a framed message from the reader
func ReadMessage(r io.Reader) (byte, []byte, error) {
	header := make([]byte, 5)
	if _, err := io.ReadFull(r, header); err != nil {
		return 0, nil, err
	}
	msgType := header[0]
	length := binary.BigEndian.Uint32(header[1:])
	if length > 16*1024*1024 { // max 16MB
		return 0, nil, fmt.Errorf("message too large: %d", length)
	}
	payload := make([]byte, length)
	if length > 0 {
		if _, err := io.ReadFull(r, payload); err != nil {
			return 0, nil, err
		}
	}
	return msgType, payload, nil
}

File: android/app/src/main/java/com/doubleagent/Protocol.kt
package com.doubleagent

object Protocol {
    const val MSG_VIDEO_INFO: Byte = 0x01
    const val MSG_VIDEO_FRAME: Byte = 0x02
    const val MSG_AUDIO_INFO: Byte = 0x03
    const val MSG_AUDIO_PACKET: Byte = 0x04
    const val MSG_TOUCH_EVENT: Byte = 0x05
    const val MSG_HEARTBEAT: Byte = 0x06

    const val TOUCH_DOWN: Byte = 0x00
    const val TOUCH_MOVE: Byte = 0x01
    const val TOUCH_UP: Byte = 0x02

    data class VideoInfo(val width: Int, val height: Int, val fps: Int)
    data class AudioInfo(val sampleRate: Int, val channels: Int, val bits: Int)
    data class TouchEvent(val x: Int, val y: Int, val action: Byte, val pointerId: Byte)
    data class TimestampedData(val timestampUs: Long, val data: ByteArray)

    fun encodeVideoInfo(v: VideoInfo): ByteArray {
        return byteArrayOf(
            (v.width shr 8).toByte(), v.width.toByte(),
            (v.height shr 8).toByte(), v.height.toByte(),
            v.fps.toByte()
        )
    }

    fun decodeVideoInfo(data: ByteArray): VideoInfo {
        require(data.size >= 5)
        val width = ((data[0].toInt() and 0xFF) shl 8) or (data[1].toInt() and 0xFF)
        val height = ((data[2].toInt() and 0xFF) shl 8) or (data[3].toInt() and 0xFF)
        val fps = data[4].toInt() and 0xFF
        return VideoInfo(width, height, fps)
    }

    fun encodeAudioInfo(a: AudioInfo): ByteArray {
        return byteArrayOf(
            (a.sampleRate shr 24).toByte(), (a.sampleRate shr 16).toByte(),
            (a.sampleRate shr 8).toByte(), a.sampleRate.toByte(),
            a.channels.toByte(), a.bits.toByte()
        )
    }

    fun decodeAudioInfo(data: ByteArray): AudioInfo {
        require(data.size >= 6)
        val sampleRate = ((data[0].toInt() and 0xFF) shl 24) or
                ((data[1].toInt() and 0xFF) shl 16) or
                ((data[2].toInt() and 0xFF) shl 8) or
                (data[3].toInt() and 0xFF)
        val channels = data[4].toInt() and 0xFF
        val bits = data[5].toInt() and 0xFF
        return AudioInfo(sampleRate, channels, bits)
    }

    fun encodeTouchEvent(t: TouchEvent): ByteArray {
        return byteArrayOf(
            (t.x shr 8).toByte(), t.x.toByte(),
            (t.y shr 8).toByte(), t.y.toByte(),
            t.action, t.pointerId
        )
    }

    fun decodeTouchEvent(data: ByteArray): TouchEvent {
        require(data.size >= 6)
        val x = ((data[0].toInt() and 0xFF) shl 8) or (data[1].toInt() and 0xFF)
        val y = ((data[2].toInt() and 0xFF) shl 8) or (data[3].toInt() and 0xFF)
        val action = data[4]
        val pointerId = data[5]
        return TouchEvent(x, y, action, pointerId)
    }

    fun extractTimestamp(data: ByteArray): TimestampedData {
        require(data.size >= 8)
        val ts = ((data[0].toLong() and 0xFF) shl 56) or
                ((data[1].toLong() and 0xFF) shl 48) or
                ((data[2].toLong() and 0xFF) shl 40) or
                ((data[3].toLong() and 0xFF) shl 32) or
                ((data[4].toLong() and 0xFF) shl 24) or
                ((data[5].toLong() and 0xFF) shl 16) or
                ((data[6].toLong() and 0xFF) shl 8) or
                (data[7].toLong() and 0xFF)
        val payload = ByteArray(data.size - 8)
        System.arraycopy(data, 8, payload, 0, data.size - 8)
        return TimestampedData(ts / 1000, payload) // ns to us
    }

    fun writeMessage(output: java.io.OutputStream, type: Byte, payload: ByteArray) {
        val frame = ByteArray(5 + payload.size)
        frame[0] = type
        frame[1] = (payload.size shr 24).toByte()
        frame[2] = (payload.size shr 16).toByte()
        frame[3] = (payload.size shr 8).toByte()
        frame[4] = payload.size.toByte()
        System.arraycopy(payload, 0, frame, 5, payload.size)
        output.write(frame)
        output.flush()
    }

    fun readMessage(input: java.io.InputStream): Pair<Byte, ByteArray>? {
        val header = ByteArray(5)
        var read = 0
        while (read < 5) {
            val n = input.read(header, read, 5 - read)
            if (n == -1) return null
            read += n
        }
        val type = header[0]
        val length = ((header[1].toInt() and 0xFF) shl 24) or
                ((header[2].toInt() and 0xFF) shl 16) or
                ((header[3].toInt() and 0xFF) shl 8) or
                (header[4].toInt() and 0xFF)
        if (length > 16 * 1024 * 1024) return null
        if (length == 0) return type to ByteArray(0)
        val payload = ByteArray(length)
        read = 0
        while (read < length) {
            val n = input.read(payload, read, length - read)
            if (n == -1) return null
            read += n
        }
        return type to payload
    }
}
